File: array/90.py
class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = []
        nums.sort()
        self.dfs(ret, nums, 0, [])
        return ret
    
    def dfs(self, ret, nums, start, temp):
        ret.append(temp)
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i-1]:
                continue
            self.dfs(ret, nums, i + 1, temp + [nums[i]])
        
        
        

File: backtracking/90.cc
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> ret;
        std::vector<int> cur;
        recur(ret, 0, nums, cur);
        return ret;
    }
private:
    void recur(std::vector<std::vector<int>>& ret, int index, std::vector<int>& nums, std::vector<int>& cur) {
        ret.emplace_back(cur);
        for (int i = index; i < nums.size(); ++i) {
            if (i != index && nums[i] == nums[i-1])
                continue;
            cur.emplace_back(nums[i]);
            recur(ret, i+1, nums, cur);
            cur.pop_back();
        }
        return;
    }
};
